### PR TITLE
CBO-300 Travel expenses: Dynamic location/destination field

### PR DESCRIPTION
**Pull request by cleargif**
#### What
AGFS:
- This feature is not active on AGFS at the moment.
- AGFS & LGFS do share the view

LGFS:
- Integrating the `establishment` data
- Introducing new travel reasons that map to particular establishments
- Managing the state and content for each expense block depending on options selected
- (while not breaking AGFS)

#### Ticket

[CBO-300](https://dsdmoj.atlassian.net/browse/CBO-300)

#### Why
- This feature is to enable the distance lookup feature [CBO-151](https://dsdmoj.atlassian.net/browse/CBO-151)

#### How
- Using a number of new `moj.Helpers` & `moj.Modules` all the moving parts have been abstracted and tested seperately
- New cucumber features test the in page states as the user makes changes

#### TODO
 - [ ] Error states  are not entirely handled properly

### CODE REVIEW

File: app/assets/javascripts/modules/Helpers.FormControls.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    optionsArray.push(new Option('Please select', '').outerHTML);
+    collectionSize = collection.length;
+
+    collection.forEach(function(obj, idx) {
+
+      var option = new Option(obj.name, obj.id, (obj[selected.prop] === selected.value));
+
+      option.dataset.postcode  = obj.postcode;
```
Comment (lostie) [inline app/assets/javascripts/modules/Helpers.FormControls.js]: **[not a blocker]** **eslint** is 👀

File: app/assets/javascripts/modules/Helpers.API.Establishments.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        error: error,
+        status: status
+      };
+
+      // Publish the success settings.events.cacheLoadError
+      $.publish(settings.events.cacheLoadError, internalCache);
+    });
+  }
+
+  // Query will merge the setttings
```
Comment (lostie) [inline app/assets/javascripts/modules/Helpers.API.Establishments.js]: **[nit pick]** misspell `setttings`

File: app/assets/javascripts/modules/Helpers.API.Establishments.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  // Filter by category
+  function getLocationByCategory(category) {
+    // If no catebory return the entire internalCache
+    if (!category) {
+      return internalCache;
+    }
+
+    // Filter the internalCache on catergory
```
Comment (lostie) [inline app/assets/javascripts/modules/Helpers.API.Establishments.js]: **[nit pick]** misspell in `catergory`

**Review by lostie — APPROVED** [review]
Pending testing in staging, the frontend changes LGTM 👍

**Comment by cleargif** [comment]
@colinbruce can you please review the Ruby changes?

**Review by colinbruce — APPROVED** [review]
Looking good :+1:

**Comment by cleargif** [comment]
See pr: #2497

**Pull request closed without merging**